File: backend/src/main/java/com/example/backend/Base64/Base64.java
package com.example.backend.Base64;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.util.LinkedHashSet;
import java.util.Set;

@Component
public class Base64 {

    private static final String DEFAULT_VALUE = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
    @Value("${spring.key}")
    private String key;
    private String createTable(){
        Set<Character> set = new LinkedHashSet<>();
        for (char ch : key.toCharArray()){
            if (!set.contains(ch)){
                set.add(ch);
            }
        }
        for (char ch : DEFAULT_VALUE.toCharArray()){
            if (!set.contains(ch)){
                set.add(ch);
            }
        }
        StringBuilder sb = new StringBuilder();
        for (char val: set){
            sb.append(val);
        }
        return sb.toString();
    }

    public String encodeData(String data){
        String table = createTable();
        String enc = java.util.Base64.getEncoder().encodeToString(data.getBytes());
        String res = "";
        for (char ch : enc.toCharArray()){
            int i = DEFAULT_VALUE.indexOf(ch);
            if (i != -1){
                res += table.charAt(i);
            }else {
                res += ch;
            }
        }
        return res;
    }

    public String decodeDate(String data){
        String table = createTable();
        String res = "";
        for (char ch : data.toCharArray()){
            int i = table.indexOf(ch);
            if (i != -1){
                res += DEFAULT_VALUE.charAt(i);
            }else {
                res += ch;
            }
        }
        byte[] dec = java.util.Base64.getDecoder().decode(res.getBytes());
        return new String(dec);
    }
}

File: webapp/src/common/Base64/Base64.js
const DEFAULT_VALUE = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
const key = "myKeySecret";

const createTable = () => {
    const set = new Set();

    for (const element of key) {
        set.add(element);
    }

    for (const element of DEFAULT_VALUE) {
        set.add(element);
    }

    return [...set].join('');
};

export const encode = (data) => {
    let ans = "";
    const encoded = btoa(data);
    const table = createTable();

    for (const ele of encoded) {
        const i = DEFAULT_VALUE.indexOf(ele);
        ans += (i !== -1) ? table.charAt(i) : ele;
    }

    return ans;
};

export const decode = (data) => {
    let ans = "";
    const table = createTable();

    const map = new Map();
    for (let i = 0; i < table.length; i++) {
        map.set(table.charAt(i), DEFAULT_VALUE.charAt(i));
    }

    for (const element of data) {
        ans += map.get(element) || element;
    }

    let decodedString;
    try {
        decodedString = atob(ans);
    } catch {
        return ans;
    }

    try {
        return JSON.parse(decodedString);
    } catch {
        return decodedString;
    }
};